### PR TITLE
Step-8: Searching Tasks

### PR DESCRIPTION
**Pull request by fza-wtag**
## Description
[Step-8](https://trello.com/c/9rq2pA9y)

This PR introduces all changes regarding implementing search.

## Checklist
- [x] On search, the users now can see tasks with a similar title to the search state.
- [x] Added Debouncing into search.
- [x] Searching works fine with filtering.
- [x] On search, pagination gets reset.
- [x] While searching, all the buttons and input elements throughout the UI gets disabled.

## Attachments
https://user-images.githubusercontent.com/119289799/217806040-767d23ea-8524-48a0-8174-6ac451033aa0.mov

### CODE REVIEW

File: todos/src/components/FilterButtons.js [hunk 1]
```diff
@@ -17,34 +17,35 @@ const FilterButtons = () => {
     dispatch(setFilter("complete"));
   };
 
+  const loadingState = useSelector(
+    (state) => state.laodingReducer.loadingState
+  );
+
+  const classNameLogicForMainDiv = `btn__progress_button ${
+    (ListLength === 0 || loadingState) && "btn__progress_button--disable"
+  }`;
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:26]:
```suggestion
  const mainDivClasses = classnames('btn__progress_button', {
  'btn__progress_button--disable': (ListLength === 0 || loadingState)
  })
```

File: todos/src/components/FilterButtons.js [hunk 2]
```diff
@@ -17,34 +17,35 @@ const FilterButtons = () => {
...
+    <div className={classNameLogicForMainDiv}>
       <button
         className={
-          filter === "all" && ListLength !== 0
-            ? "btn__progress_button--disable"
-            : null
+          filter === "all" &&
+          ListLength !== 0 &&
+          !loadingState &&
+          "btn__progress_button--disable"
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:35]: Let's use computed properties here

File: todos/src/components/FilterButtons.js [hunk 3]
```diff
@@ -17,34 +17,35 @@ const FilterButtons = () => {
...
         }
         onClick={handleAllTodos}
       >
         All
       </button>
       <button
-        className={
-          filter === "incomplete" ? "btn__progress_button--disable" : null
-        }
+        className={filter === "incomplete" && "btn__progress_button--disable"}
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:42]: Let's use constants here

File: todos/src/components/FilterButtons.js [hunk 4]
```diff
@@ -17,34 +17,35 @@ const FilterButtons = () => {
...
         onClick={handleIncompleteTodos}
       >
         Incomplete
       </button>
       <button
-        className={
-          filter === "complete" ? "btn__progress_button--disable" : null
-        }
+        className={filter === "complete" && "btn__progress_button--disable"}
```
Comment (hhu-wtag) [inline todos/src/components/FilterButtons.js:48]: ditto

File: todos/src/components/Navbar.js [hunk 5]
```diff
@@ -1,17 +1,59 @@
...
+      dispatch(setLoadingState(true));
+      if (timer) clearTimeout(timer);
+      timer = setTimeout(() => {
+        timer = null;
+        func.apply(context, args);
+        dispatch(setLoadingState(false));
+      }, 1000);
+      
+    };
+  };
```
Comment (hhu-wtag) [inline todos/src/components/Navbar.js:38]: Let's generalize the debounce function and move it to `utils.js` file for modularity.
```suggestion
  // utils.js
  export const debounce = (func, delay) => {
    let timer;
    return function (...args) {
      const context = this;
      if (timer) clearTimeout(timer);
      timer = setTimeout(() => {
        timer = null;
        func.apply(context, args);
      }, delay);
    };
  };
```

File: todos/src/components/Todos.js [hunk 6]
```diff
@@ -16,20 +17,32 @@ function Todos() {
...
+  const searchedTodos = filteredTodos.filter((elem) =>
+    elem.data.toLowerCase().includes(searchValue.toLowerCase())
+  );
+  const displayedTodoList = searchedTodos.slice(0, PER_PAGE * currentPage);
+  const loadingState = useSelector(
+    (state) => state.laodingReducer.loadingState
+  );
+
+  const showEmptyListIcon = displayedTodoList.length === 0 && !isAddTaskVisible;
+  const lessThanfilteredTodos = currentPage * PER_PAGE < filteredTodos.length;
```
Comment (hhu-wtag) [inline todos/src/components/Todos.js:37]:
```suggestion
  const lessThanFilteredTodos = currentPage * PER_PAGE < filteredTodos.length;
```

File: todos/src/components/Todos.js [hunk 7]
```diff
@@ -16,20 +17,32 @@ function Todos() {
...
+  const lessThansearchedTodos = currentPage * PER_PAGE < searchedTodos.length;
```
Comment (hhu-wtag) [inline todos/src/components/Todos.js:38]:
```suggestion
  const lessThanSearchedTodos = currentPage * PER_PAGE < searchedTodos.length;
```

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: todos/src/components/Navbar.js [hunk 8]
```diff
@@ -1,17 +1,40 @@
...
+  const searchIconState = useSelector((state) => state.searchReducer.iconState);
+  const handleSearchIconClick = () => {
+    dispatch(setIconVisibility(!searchIconState));
+  };
+
+  const handleSearchInput = (event) => {
+    dispatch(setSearchValue(event.target.value));
+  };
+
+  const optimizedHandle = useCallback(debounce(handleSearchInput), []);
```
Comment (hhu-wtag) [inline todos/src/components/Navbar.js:21]:
```suggestion
  const handleSearch = useCallback(debounce(handleSearchInput), []);
```

File: todos/src/components/Navbar.js [hunk 9]
```diff
@@ -1,17 +1,40 @@
...
         <h4>Todos</h4>
       </div>
       <div className="navbar__search">
-        <input type="search" placeholder="Search" />
-        <button type="submit">
+        {!searchIconState && (
+          <input
+            type="search"
+            placeholder="Search"
+            onChange={optimizedHandle}
```
Comment (hhu-wtag) [inline todos/src/components/Navbar.js:34]:
```suggestion
            onChange={handleSearch}
```

File: todos/src/components/Todos.js [hunk 10]
```diff
@@ -48,13 +56,23 @@ function Todos() {
...
         })}
       </div>
+      {loadingState && (
+        <img className="spinner" src={spinner} alt="Loging"></img>
+      )}
       {showEmptyListIcon && <EmptyTaskList />}
-      {lessThanListLength ? (
+      {searchValue.length === 0 ? (
+        lessThanFilteredTodos ? (
+          <LoadMoreBtn type={LOAD_MORE} />
```
Comment (hhu-wtag) [inline todos/src/components/Todos.js:68]: As we are using the same button for two different purpose, we should name it more generally. 
E.G: `PaginatorButton`

File: todos/src/components/Todos.js [hunk 11]
```diff
@@ -48,13 +56,23 @@ function Todos() {
...
+        ) : (
+          filteredTodos.length > PER_PAGE && <LoadMoreBtn type={SHOW_LESS} />
+        )
+      ) : lessThanSearchedTodos ? (
```
Comment (hhu-wtag) [inline todos/src/components/Todos.js:72]: This looks too complex at first sight. Can we simplify this?

File: todos/src/reducers/loadingReducer.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+const initialState = {
+  loadingState: false,
+};
+const laodingReducer = (state = initialState, action) => {
```
Comment (hhu-wtag) [inline todos/src/reducers/loadingReducer.js:4]: typo
```suggestion
const loadingReducer = (state = initialState, action) => {
```

File: todos/src/utils.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import store from "./store";
+import { setLoadingState } from "actions";
+
+export const debounce = (func) => {
+  let timer;
+
+  return function (...args) {
+    const context = this;
+    store.dispatch(setLoadingState(true));
```
Comment (hhu-wtag) [inline todos/src/utils.js:9]: Can we remove the dispatch from here? We want this function to be as generic as possible. 

We can add the loading function in the context of where debounce is used.

**Review by hhu-wtag — CHANGES_REQUESTED** [review]

File: todos/src/components/Todos.js (hunk 11 above)
Comment (fza-wtag) [reply todos/src/components/Todos.js:72]: Sure @hhu-wtag Bhai we can. Doing it right now.

File: todos/src/reducers/searchReducer.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+const initialState = {
+  iconState: "false",
```
Comment (hhu-wtag) [inline todos/src/reducers/searchReducer.js:2]: is **false** use as string here? Seems counter intuitive.

**Review by hhu-wtag — APPROVED** [review]

**Review by ray-wtag — APPROVED** [review]

**Pull request merged**